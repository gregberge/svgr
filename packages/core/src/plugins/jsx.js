import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
