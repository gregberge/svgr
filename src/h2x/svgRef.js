import { JSXAttribute } from 'h2x-plugin-jsx'

const svgRef = () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (
          path.node.name === 'svg' &&
          !path.node.attributes.some(attr => attr && attr.name === 'ref')
        ) {
          const props = new JSXAttribute()
          props.name = 'ref'
          props.value = 'svgRef'
          props.litteral = true
          path.node.attributes.push(props)
          path.replace(path.node)
        }
      },
    },
  },
})

export default svgRef
