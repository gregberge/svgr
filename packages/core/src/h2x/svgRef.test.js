import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import svgRef from './svgRef'

describe('svgRef', () => {
  it('should add a svg ref attribute', () => {
    const result = transform(
      `<svg width="20" height="20">
        <g stroke="#063855" stroke-width="2" />
      </svg>`,
      { plugins: [jsx, svgRef()] },
    )

    expect(result).toMatchSnapshot()
  })
})
