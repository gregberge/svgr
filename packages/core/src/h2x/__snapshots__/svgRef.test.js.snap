// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgRef should add a svg ref attribute 1`] = `
"<svg width={20} height={20} ref={svgRef}>
  <g stroke=\\"#063855\\" strokeWidth={2} />
</svg>
"
`;
